hw02-gradle/src/main/java/ru/otus/DIYarrayList.java

```diff
@@ -0,0 +1,334 @@
+package ru.otus;
+
+import java.util.*;
+
+public class DIYarrayList<T> implements List<T> {
+    public static final int MAX_ARRAY_LENGTH = Integer.MAX_VALUE - 8;
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private Object[] elementData;
+    private int size;
+
+    public DIYarrayList(int initialCapacity) {
+        if (initialCapacity > 0) {
+            try {
+                this.elementData = new Object[initialCapacity];
+            }catch (ClassCastException e){
+                e.printStackTrace();
+            }
+        } else {
+            throw new IllegalArgumentException("Illegal Capacity: " + initialCapacity);
+        }
+    }
+
+    public DIYarrayList() {
+        this.elementData = new Object[DEFAULT_CAPACITY];
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        return new DIYarrayList.Itr();
+    }
+
+    @Override
+    public Object[] toArray() {
+        return Arrays.copyOf(elementData, size);
+    }
+
+    @Override
+    public <T1> T1[] toArray(T1[] a) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean add(T t) {
+        if (size == elementData.length) {
+            Object[] bufList = elementData;
+            if(size == MAX_ARRAY_LENGTH){
+                throw new OutOfMemoryError("Required array length too large");
```

Comment:
Жестоко сразу Error кидать. Error мы не кидаем в прикладных приложениях, это системая фича. Он сразу рушит всё приложение

Reply:
Подсмотрел это поведение в методе ArraysSupport::hugeLength, который вызывается в ArrayList::grow, когда newLength > MAX_ARRAY_LENGTH.
Исправил, на throw new IllegalStateException(). Как я понял, просто return true не подойдет, т.к. в Collections::add написано следующее:
     * If a collection refuses to add a particular element for any reason
     * other than that it already contains the element, it <i>must</i> throw
     * an exception (rather than returning {@code false}).  This preserves
     * the invariant that a collection always contains the specified element
     * after this call returns.
Там же прочитал:
     * @throws IllegalStateException if the element cannot be added at this
     *         time due to insertion restrictions
Вроде как, это наиболее подходящая ситуация.
Правильно ли я понял?